gtk4/src/tree_iter.rs

```diff
@@ -0,0 +1,152 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+use cairo::glib::Object;
+use glib::ffi::gpointer;
+use glib::object::GObject;
+use glib::translate::*;
+use glib::ObjectType;
+
+use std::mem;
+use std::os::raw::c_int;
+
+// 0xffffffff is an invalid pointer that is distinct from the null pointer
+// This makes it unlikely this is in here by accident
+const MAGIC_NUMBER_1: usize = 0xffffffff;
+// "rust" in ASCII
+const MAGIC_NUMBER_2: usize = 0x72757374;
+
+#[repr(C)]
+#[derive(Clone)]
+#[doc(alias = "GtkTreeIter")]
+pub struct TreeIter {
+    pub stamp: c_int,
+    pub user_data: gpointer,
+    pub user_data2: gpointer,
+    pub user_data3: gpointer,
+}
+
+impl TreeIter {
+    /// Create a TreeIter referencing a glib::Object
+    pub fn new(stamp: i32, item: Object) -> Self {
+        assert_initialized_main_thread!();
+
+        // We only really need the first user_data field for our data.
+        // We will store a reference to the glib object here.
+        let user_data = unsafe { glib::gobject_ffi::g_object_ref(item.as_ptr()) } as gpointer;
```

Comment:
This would leak the object if the iter is passed to C and freed there. Also if you clone the iter you wouldn't increase the refcount of the object, but then unref it twice later.

You can't store any owned memory in an iter, the API does not really allow such things. But for safely storing unowned/borrowed pointers you need something else.